lib/kasan/kasan.c

```diff
@@ -0,0 +1,340 @@
+
+	if (likely(shadow_val == 0 || last < shadow_val))
+		return true;
+	*code = shadow_val;
+	return false;
+}
+
+static inline bool
+shadow_4byte_isvalid(uintptr_t addr, uint8_t *code) {
+	if (!access_within_shadow_byte(addr, 4))
+		return shadow_2byte_isvalid(addr, code) &&
+			shadow_2byte_isvalid(addr + 2, code);
+
+	int8_t shadow_val = *kasan_md_addr_to_shad(addr);
+	int8_t last = (addr + 3) & KASAN_SHADOW_MASK;
+
+	if (likely(shadow_val == 0 || last < shadow_val))
+		return true;
+	*code = shadow_val;
+	return false;
+}
+
+static inline bool
+shadow_8byte_isvalid(uintptr_t addr, uint8_t *code)
+{
+	if (!access_within_shadow_byte(addr, 8))
+		return shadow_4byte_isvalid(addr, code) &&
+			shadow_4byte_isvalid(addr + 4, code);
+
+	int8_t shadow_val = *kasan_md_addr_to_shad(addr);
+	int8_t last = (addr + 7) & KASAN_SHADOW_MASK;
+
+	if (likely(shadow_val == 0 || last < shadow_val))
+		return true;
+	*code = shadow_val;
+	return false;
+}
+
+static inline bool
+shadow_Nbyte_isvalid(uintptr_t addr, size_t size, uint8_t *code)
+{
+		for (size_t i = 0; i < size; i++)
+			if (unlikely(!shadow_1byte_isvalid(addr + i, code)))
+				return false;
+		return true;
+}
+
+static inline void
+shadow_check(uintptr_t addr, size_t size, bool read)
+{
+	if (unlikely(!kasan_ready))
+		return;
+	if (unlikely(!kasan_md_addr_supported(addr)))
+		return;
+
+	uint8_t code = 0;
+	bool valid = true;
+
+	if (__builtin_constant_p(size)) {
+		switch (size) {
+		case 1:
+			valid = shadow_1byte_isvalid(addr, &code);
+			break;
+		case 2:
+			valid = shadow_2byte_isvalid(addr, &code);
+			break;
+		case 4:
+			valid = shadow_4byte_isvalid(addr, &code);
+			break;
+		case 8:
+			valid = shadow_8byte_isvalid(addr, &code);
+			break;
+		}
+	} else {
+		valid = shadow_Nbyte_isvalid(addr, size, &code);
+	}
+
+	if (unlikely(!valid)) {
+		UK_CRASH("===========KernelAddressSanitizer===========\n"
+			"ERROR:\n"
+			"* invalid access to address %p\n"
+			"* %s of size %lu\n"
+			"* redzone code 0x%x (%s)\n"
+			"============================================\n",
+			(void *)addr, (read ? "read" : "write"), size, code,
+			code_name(code));
+	}
+}
+
+/*
+ * Memory is divided into 8-byte blocks aligned to 8-byte boundary. Each block
+ * has corresponding descriptor byte in the shadow memory. You can mark each
+ * block as valid (0x00) or invalid (0xF1 - 0xFF). Blocks can be partially valid
+ * (0x01 - 0x07) - i.e. prefix is valid, suffix is invalid.  Other variants are
+ * NOT POSSIBLE! Thus `addr` and `total` must be block aligned.
+ */
+void kasan_mark(const void *addr, size_t valid, size_t total, uint8_t code)
+{
+	UK_ASSERT(is_aligned(addr, KASAN_SHADOW_SCALE_SIZE));
+	UK_ASSERT(is_aligned(total, KASAN_SHADOW_SCALE_SIZE));
+	UK_ASSERT(valid <= total);
+
+	int8_t *shadow = kasan_md_addr_to_shad((uintptr_t)addr);
+	int8_t *end = shadow + total / KASAN_SHADOW_SCALE_SIZE;
+
+	/* Valid bytes. */
+	size_t len = valid / KASAN_SHADOW_SCALE_SIZE;
+
+	__builtin_memset(shadow, 0, len);
```

Comment:
Is there a reason to explicitly use `__builtin_memset`? It can fallback to a regular `memset` call (and when I tried this PR, it also did exactly that!).

Reply:
Nothign special, I've actually reverted to using memset.